mmv1/third_party/tgc_next/test/hcl.go

```diff
@@ -94,27 +96,69 @@ func insert(data any, key string, parent map[string]any) {
 	}
 }
 
-func flatten(data interface{}, prefix string, result map[string]struct{}) {
+func flatten(data any, prefix string, result map[string]struct{}) {
 	switch v := data.(type) {
-	case map[string]interface{}:
+	case map[string]any:
 		for key, value := range v {
 			newPrefix := key
 			if prefix != "" {
 				newPrefix = prefix + "." + key
 			}
 			flatten(value, newPrefix, result)
 		}
-	case []interface{}:
-		if len(v) == 0 && prefix != "" {
-			result[prefix] = struct{}{}
-		}
-		for i, value := range v {
-			newPrefix := fmt.Sprintf("%s.%d", prefix, i)
-			flatten(value, newPrefix, result)
-		}
+	case []any:
+		flattenSlice(prefix, v, result)
 	default:
 		if prefix != "" {
 			result[prefix] = struct{}{}
 		}
 	}
 }
+
+func flattenSlice(prefix string, v []any, result map[string]struct{}) {
+	if len(v) == 0 && prefix != "" {
+		result[prefix] = struct{}{}
+		return
+	}
+
+	type sortableElement struct {
+		flatKeys  string
+		flattened map[string]struct{}
+	}
+
+	sortable := make([]sortableElement, len(v))
+	for i, value := range v {
```

Comment:
If I understood correctly, this function sorts top level elements with the same field name by checking the sub-fields keys, and sort them in alphabetical order?

For example:
allow {
   test: "test"
}

allow {
  "port": "port"
}

into allow {"port": "port", "test", "test"}

Reply:
Yes, it will sort and flatten them into `allow.0.port` and `allow.1.test`